### PR TITLE
Extend -o to work when compiling C files

### PR DESCRIPTION
**Pull request by shindere**
This PR fixes issue #7677.

When using ocamlc/ocamlopt to compile C files, it is currently not possible
to control where the C compiler should place the resulting object file.

In other words, without this PR, a command like

```shell
ocamlopt -c foo.c -o /tmp/foo.o
```

produced the following error message:

Options -c and -o are incompatible when compiling C files

This PR removes this restriction and lets the compiler accept
that the command-line options -c and -o are given simultaneously
also when compiling one C file, so that the previous command now works.

The -o command-line option remains however forbidden when compiling
several C files during the same compiler invocation.

The PR includes tests to ensure the compiler behaves as expected when
only one C file is provided on the command-line and when several
of them are provided.

A `file-exists` action has been added to ocamltest to make testing for the
presence of the object file easier.

The code has been tested with precheck, see build
[##521](https://ci.inria.fr/ocaml/job/precheck/521/).

Cc @dbuenzli

### CODE REVIEW

File: driver/compenv.ml [hunk 1]
```diff
@@ -656,12 +661,12 @@ let process_deferred_actions env =
...
+    | Some _output_name ->
         if !compile_only then begin
-          if List.filter (function
-              | ProcessCFile name -> c_object_of_filename name <> output_name
-              | _ -> false) !deferred_actions <> [] then
-            fatal "Options -c and -o are incompatible when compiling C files";
+          if List.length (List.filter (function
+              | ProcessCFile _ -> true
+              | _ -> false) !deferred_actions) > 1 then
+            fatal "Option -o makes no sense when compiling several C files";
```
Comment (Octachron) [inline driver/compenv.ml]: Shouldn't this case should be combined with the one below to reject
```ocaml
ocamlopt -c a.c b.ml -o something
```
?

**Comment by shindere** [comment]
Florian Angeletti (2020/10/13 01:16 -0700):
> Shouldn't this case should be combined with the one below to reject
> ```ocaml
> ocamlopt -c a.c b.ml -o something
> ```
> ?

That was definitely a good idea, many thanks for the suggestion.

The PR has been rebased on latest trunk and updated accordingly.

**Comment by shindere** [comment]
Is there anything that prevents this from being approved / merged?

Would be happy to fix any remaining issue.

**Review by gasche — APPROVED** [review]
The code looks good and the feature is useful.

The change needs a Changes entry.

**Comment by shindere** [comment]
Many thanks for your prompt response and approval, @gasche.

The PR has been rebased on latest trunk and a Changes entry added.

**Comment by gasche** [comment]
Thanks! `check-typo` complains:

```
Checking 6f76c388bee0b478dbbb2c2d82e46b389fae66f8: ocamltest/builtin_actions.ml
./ocamltest/builtin_actions.ml:224.35: [white-at-eol] whitespace at end of line
./ocamltest/builtin_actions.ml:226.15: [white-at-eol] whitespace at end of line
```

**Comment by shindere** [comment]
Gabriel Scherer (2021/01/05 04:18 -0800):
> Thanks! `check-typo` complains:
[...]

Gosh. Sorry, and thanks. Fixed.

**Comment by gasche** [comment]
`check-typo` is still complaining in the same way, are you sure that you pushed your fixing change?

**Comment by shindere** [comment]
Gabriel Scherer (2021/01/05 08:48 -0800):
> `check-typo` is still complaining in the same way, are you sure that
> you pushed your fixing change?

They should be there now. Sorry.

**Comment by shindere** [comment]
Gabriel Scherer (2021/01/06 00:07 -0800):
> Merged #9960 into trunk.

Great! Thanks!

**Comment by nojb** [comment]
There are two issues with this patch:

1. It broke previous behaviour of `ocaml` when you *don't* pass `-o`:
```
(before) ocamlc foo.c --> calls "$cc -c foo.c"
(after) ocamlc foo.c --> calls "$cc -o foo.o -c foo.c"
```
This causes problems when people use:
```
ocamlc -ccopt '-o whatever' foo.c
```
which was necessary before this PR.

2. The `-o` flag to `ocaml{c,opt}` is ignored if `-c` is not passed as well when compiling C files.

**Comment by ghost** [comment]
cf. https://github.com/moosotc/llpp/issues/174

**Pull request merged**